adobe/copy_on_write.hpp

```diff
@@ -55,93 +36,111 @@ copy when the value is modified and there is more than one reference to the
-        instance of value_type which will be used for subsequent calls to the
+        instance of element_type which will be used for subsequent calls to the
         default constructor. The default instance will be released at exit.
     */
     copy_on_write() {
-        std::call_once(flag_s, init_default);
-        object_m = default_s;
-        object_m->header_m.get().count_m.increment();
+        static model default_s;
+        _self = &default_s;
+        ++_self->_count;
     }
 
     /*!
-        Initializes the instance with a custom allocator
+    \brief Constructs a new copy_on_write object with a single argument.
     */
-    explicit copy_on_write(const allocator_type& a) : object_m(0) {
-        implementation_allocator_type other_allocator(a);
-
-        object_m = allocate(other_allocator);
-    }
+    template <class U>
+    copy_on_write(U&& x, disable_copy<U> = nullptr) : _self(new model(std::forward<U>(x))) {}
 
     /*!
-    \brief Constructs a new copy_on_write object with a value \c x.
-
-    \param x A default value to assign to this object
+    \brief Constructs a new copy_on_write object with two or more arguments.
     */
-    copy_on_write(T x) : object_m(allocate_move(0, std::move(x))) {}
+    template <class U, class V, class... Args>
+    copy_on_write(U&& x, V&& y, Args&&... args)
+        : _self(new model(std::forward<U>(x), std::forward<V>(y), std::forward<Args>(args)...)) {}
 
     /*!
     Copy construction is a non-throwing operation and simply increments the
     reference count on the stored object.
     */
-    copy_on_write(const copy_on_write& x) : object_m(x.object_m) {
-        if (object_m)
-            object_m->header_m.get().count_m.increment();
-    }
+    copy_on_write(const copy_on_write& x) noexcept : _self(x._self) {
+        assert(_self && "FATAL (sparent) : using a moved copy_on_write object");
 
-    copy_on_write(copy_on_write&& x) : object_m(x.object_m) { x.object_m = 0; }
+        ++_self->_count;
+    }
+    copy_on_write(copy_on_write&& x) noexcept : _self(x._self) {
+        assert(_self && "WARNING (sparent) : using a moved copy_on_write object");
+        x._self = nullptr;
+    }
 
-    ~copy_on_write() { release(object_m); }
+    ~copy_on_write() { destroy(); }
 
     /*!
         As with copy construction, assignment is a non-throwing operation which
         releases the old value and increments the reference count of the item
         being assigned to.
     */
-    copy_on_write& operator=(copy_on_write x) {
-        swap(*this, x);
-        return *this;
+    auto operator=(const copy_on_write& x) noexcept -> copy_on_write& {
+        return *this = copy_on_write(x);
     }
+    auto operator=(copy_on_write&& x) noexcept -> copy_on_write& {
+        assert(x._self && "WARNING (sparent) : using a moved copy_on_write object");
+        assert((&x != this) && "FATAL (sparent) : invalid self-move.");
```

Comment:
This will also make self-copy-assignment invalid, because the copy-ctor runs through this routine. Is that lumped into the same conversation as self-move? I'm not sure that it is. Consider:

```
copy_on_write<int> x(42);
copy_on_write<int> z;
z = std::min(z, x); // assert - invalid self-move
```